d-sandbox
<div style="text-align: center; line-height: 0; padding-top: 9px;">
  <img src="https://databricks.com/wp-content/uploads/2018/03/db-academy-rgb-1200px.png" alt="Databricks Learning" style="width: 400px">
</div>

# Spark SQL
1. Run a SQL query
1. Create a DataFrame
1. Write same query using DataFrame transformations
1. Trigger computation with DataFrame actions
1. Convert between DataFrames and SQL

##### Methods
- SparkSession (<a href="https://spark.apache.org/docs/latest/api/python/pyspark.sql.html?highlight=sparksession#pyspark.sql.SparkSession" target="_blank">Python</a>/<a href="http://spark.apache.org/docs/latest/api/scala/org/apache/spark/sql/SparkSession.html" target="_blank">Scala</a>): `sql`, `table`
- DataFrame (<a href="https://spark.apache.org/docs/latest/api/python/pyspark.sql.html?highlight=dataframe#pyspark.sql.DataFrame" target="_blank">Python</a>/<a href="http://spark.apache.org/docs/latest/api/scala/org/apache/spark/sql/Dataset.html" target="_blank">Scala</a>):
  - Transformations:  `select`, `where`, `orderBy`
  - Actions: `show`, `count`, `take`
  - Other methods: `printSchema`, `schema`, `createOrReplaceTempView`

In [0]:
%run ./Includes/Classroom-Setup-SQL

Datasets mounted and student environment set up

### ![Spark Logo Tiny](https://files.training.databricks.com/images/105/logo_spark_tiny.png) Run a SQL query
Use `SparkSession` to run SQL

In [0]:
budgetDF = spark.sql("""
SELECT name, price
FROM products
WHERE price < 200
ORDER BY price
""")

View results in the returned DataFrame

In [0]:
budgetDF.show()

+--------------------+-----+
 name|price|
+--------------------+-----+
Standard Foam Pillow| 59.0|
 King Foam Pillow| 79.0|
Standard Down Pillow|119.0|
 King Down Pillow|159.0|
+--------------------+-----+

In [0]:
budgetDF.toPandas()

Out[7]:

name 
 price 
 
 
 
 
 0 
 Standard Foam Pillow 
 59.0 
 
 
 1 
 King Foam Pillow 
 79.0 
 
 
 2 
 Standard Down Pillow 
 119.0 
 
 
 3 
 King Down Pillow 
 159.0

In [0]:
display(budgetDF)

name price Standard Foam Pillow 59.0 King Foam Pillow 79.0 Standard Down Pillow 119.0 King Down Pillow 159.0

### ![Spark Logo Tiny](https://files.training.databricks.com/images/105/logo_spark_tiny.png) Create a DataFrame
Use `SparkSession` to create a DataFrame from a table

In [0]:
productsDF = spark.table("products")
display(productsDF)

item_id name price M_PREM_Q Premium Queen Mattress 1795.0 M_STAN_F Standard Full Mattress 945.0 M_PREM_F Premium Full Mattress 1695.0 M_PREM_T Premium Twin Mattress 1095.0 M_PREM_K Premium King Mattress 1995.0 P_DOWN_S Standard Down Pillow 119.0 M_STAN_Q Standard Queen Mattress 1045.0 M_STAN_K Standard King Mattress 1195.0 M_STAN_T Standard Twin Mattress 595.0 P_FOAM_S Standard Foam Pillow 59.0 P_FOAM_K King Foam Pillow 79.0 P_DOWN_K King Down Pillow 159.0

Access schema of DataFrame

In [0]:
productsDF.printSchema()

root
-- item_id: string (nullable = true)
-- name: string (nullable = true)
-- price: double (nullable = true)

In [0]:
productsDF.schema

Out[11]: StructType(List(StructField(item_id,StringType,true),StructField(name,StringType,true),StructField(price,DoubleType,true)))

### ![Spark Logo Tiny](https://files.training.databricks.com/images/105/logo_spark_tiny.png) Write same query with DataFrame transformations

In [0]:
budgetDF = (productsDF
  .select("name", "price")
  .where("price < 200")
  .orderBy("price")
)

### ![Spark Logo Tiny](https://files.training.databricks.com/images/105/logo_spark_tiny.png) Trigger computation with DataFrame actions

In [0]:
budgetDF.count()

Out[13]: 4

In [0]:
budgetDF.take(2)

Out[14]: [Row(name='Standard Foam Pillow', price=59.0),
 Row(name='King Foam Pillow', price=79.0)]

### ![Spark Logo Tiny](https://files.training.databricks.com/images/105/logo_spark_tiny.png) Convert between DataFrames and SQL

In [0]:
budgetDF.createOrReplaceTempView("budget")

In [0]:
display(spark.sql("SELECT * FROM budget"))

name price Standard Foam Pillow 59.0 King Foam Pillow 79.0 Standard Down Pillow 119.0 King Down Pillow 159.0

## ![Spark Logo Tiny](https://files.training.databricks.com/images/105/logo_spark_tiny.png) Lab

1. Create a DataFrame from the `Event` table
1. Display DataFrame and inspect schema
1. Apply transformations to filter and sort `macOS` events
1. Count results and take first 5 rows
1. Create the same DataFrame using SQL query

##### Methods
- <a href="http://spark.apache.org/docs/latest/api/scala/org/apache/spark/sql/SparkSession.html" target="_blank">SparkSession</a>: `sql`, `table`
- <a href="http://spark.apache.org/docs/latest/api/scala/org/apache/spark/sql/Dataset.html" target="_blank">DataFrame</a> transformations: `select`, `where`, `orderBy`
- <a href="http://spark.apache.org/docs/latest/api/scala/org/apache/spark/sql/Dataset.html" target="_blank">DataFrame</a> actions: `select`, `count`, `take`
- Other <a href="http://spark.apache.org/docs/latest/api/scala/org/apache/spark/sql/Dataset.html" target="_blank">DataFrame</a> methods: `printSchema`, `schema`, `createOrReplaceTempView`

### 1. Create a DataFrame from the `events` table
- Use SparkSession to create a DataFrame from the `events` table

In [0]:
eventDF = spark.table('events')

### 2. Display DataFrame and inspect schema
- Use methods above to inspect DataFrame contents and schema

In [0]:
# Print events table schema
eventDF.printSchema()

root
-- device: string (nullable = true)
-- ecommerce: struct (nullable = true)
 |-- purchase_revenue_in_usd: double (nullable = true)
 |-- total_item_quantity: long (nullable = true)
 |-- unique_items: long (nullable = true)
-- event_name: string (nullable = true)
-- event_previous_timestamp: long (nullable = true)
-- event_timestamp: long (nullable = true)
-- geo: struct (nullable = true)
 |-- city: string (nullable = true)
 |-- state: string (nullable = true)
-- items: array (nullable = true)
 |-- element: struct (containsNull = true)
 | |-- coupon: string (nullable = true)
 | |-- item_id: string (nullable = true)
 | |-- item_name: string (nullable = true)
 | |-- item_revenue_in_usd: double (nullable = true)
 | |-- price_in_usd: double (nullable = true)
 | |-- quantity: long (nullable = true)
-- traffic_source: string (nullable = true)
-- user_first_touch_timestamp: long (nullable = true)
-- user_id: string (nullable = true)

In [0]:
# Printing out the binary representation of the schema
eventDF.schema

Out[23]: StructType(List(StructField(device,StringType,true),StructField(ecommerce,StructType(List(StructField(purchase_revenue_in_usd,DoubleType,true),StructField(total_item_quantity,LongType,true),StructField(unique_items,LongType,true))),true),StructField(event_name,StringType,true),StructField(event_previous_timestamp,LongType,true),StructField(event_timestamp,LongType,true),StructField(geo,StructType(List(StructField(city,StringType,true),StructField(state,StringType,true))),true),StructField(items,ArrayType(StructType(List(StructField(coupon,StringType,true),StructField(item_id,StringType,true),StructField(item_name,StringType,true),StructField(item_revenue_in_usd,DoubleType,true),StructField(price_in_usd,DoubleType,true),StructField(quantity,LongType,true))),true),true),StructField(traffic_source,StringType,true),StructField(user_first_touch_timestamp,LongType,true),StructField(user_id,StringType,true)))

-sandbox
### 3. Apply transformations to filter and sort `macOS` events
- Filter for rows where `device` is `macOS`
- Sort rows by `event_timestamp`

<img alt="Hint" title="Hint" style="vertical-align: text-bottom; position: relative; height:1.75em; top:0.3em" src="https://files.training.databricks.com/static/images/icon-light-bulb.svg"/>&nbsp;**Hint:** Use single and double quotes in your filter SQL expression

In [0]:
eventDF.orderBy?

In [0]:
# TODO
macDF = (eventDF
         .filter(eventDF.device == 'macOS')
         .orderBy("event_timestamp")
)

### 4. Count results and take first 5 rows
- Use DataFrame actions to count and take rows

In [0]:
# TODO
numRows = macDF.count()
rows = macDF.take(5)
print(numRows)
rows

1938215
Out[27]: [Row(device='macOS', ecommerce=Row(purchase_revenue_in_usd=None, total_item_quantity=None, unique_items=None), event_name='mattresses', event_previous_timestamp=None, event_timestamp=1592539216262230, geo=Row(city='Waterbury', state='CT'), items=[], traffic_source='direct', user_first_touch_timestamp=1592539216262230, user_id='UA000000103314644'),
 Row(device='macOS', ecommerce=Row(purchase_revenue_in_usd=None, total_item_quantity=None, unique_items=None), event_name='mattresses', event_previous_timestamp=1592322041626307, event_timestamp=1592539218667144, geo=Row(city='Cedar Rapids', state='IA'), items=[], traffic_source='email', user_first_touch_timestamp=1592321551374513, user_id='UA000000102690062'),
 Row(device='macOS', ecommerce=Row(purchase_revenue_in_usd=None, total_item_quantity=None, unique_items=None), event_name='reviews', event_previous_timestamp=1592538997090958, event_timestamp=1592539218939615, geo=Row(city='Toledo', state='OH'), items=[], traffic_source='google', user_first_touch_timestamp=1592538965425275, user_id='UA000000103314591'),
 Row(device='macOS', ecommerce=Row(purchase_revenue_in_usd=None, total_item_quantity=None, unique_items=None), event_name='guest', event_previous_timestamp=1592538390245546, event_timestamp=1592539225913937, geo=Row(city='Los Angeles', state='CA'), items=[Row(coupon=None, item_id='M_STAN_F', item_name='Standard Full Mattress', item_revenue_in_usd=945.0, price_in_usd=945.0, quantity=1)], traffic_source='facebook', user_first_touch_timestamp=1592535034918745, user_id='UA000000103313292'),
 Row(device='macOS', ecommerce=Row(purchase_revenue_in_usd=None, total_item_quantity=None, unique_items=None), event_name='main', event_previous_timestamp=None, event_timestamp=1592539226602157, geo=Row(city='Tulsa', state='OK'), items=[], traffic_source='google', user_first_touch_timestamp=1592539226602157, user_id='UA000000103314648')]

-sandbox
##### <img alt="Best Practice" title="Best Practice" style="vertical-align: text-bottom; position: relative; height:1.75em; top:0.3em" src="https://files.training.databricks.com/static/images/icon-blue-ribbon.svg"/> Check your work

In [0]:
from pyspark.sql import Row

assert(numRows == 1938215)
assert(len(rows) == 5)
assert(type(rows[0]) == Row)

### 5. Create the same DataFrame using SQL query
- Use SparkSession to run a sql query on the `events` table
- Use SQL commands above to write the same filter and sort query used earlier

In [0]:
eventDF.createOrReplaceTempView('events')
macSQLDF = spark.sql("""
        SELECT device, event_timestamp
        FROM events
        WHERE device = 'macOS'
        ORDER BY event_timestamp
""")

display(macSQLDF)

device event_timestamp macOS 1592539216262230 macOS 1592539218667144 macOS 1592539218939615 macOS 1592539225913937 macOS 1592539226602157 macOS 1592539229480664 macOS 1592539231022266 macOS 1592539231776200 macOS 1592539232284730 macOS 1592539233151229 macOS 1592539235059340 macOS 1592539236295422 macOS 1592539237432046 macOS 1592539289834596 macOS 1592539295857466 macOS 1592539299325832 macOS 1592539302562339 macOS 1592539305168803 macOS 1592539309500570 macOS 1592539310482088 macOS 1592539331750583 macOS 1592539339785249 macOS 1592539343571171 macOS 1592539346041388 macOS 1592539347745795 macOS 1592539367676338 macOS 1592539383618319 macOS 1592539387295729 macOS 1592539388476666 macOS 1592539388713269 macOS 1592539389057282 macOS 1592539394685322 macOS 1592539400000826 macOS 1592539402672930 macOS 1592539414743914 macOS 1592539417279950 macOS 1592539417363593 macOS 1592539418536140 macOS 1592539421435621 macOS 1592539425024905 macOS 1592539430046139 macOS 1592539430510397 macOS 1592539433623588 macOS 1592539439299177 macOS 1592539444746132 macOS 1592539446975318 macOS 1592539453152225 macOS 1592539453911179 macOS 1592539470197362 macOS 1592539476111909 macOS 1592539485558893 macOS 1592539493327941 macOS 1592539494355700 macOS 1592539495439176 macOS 1592539504294974 macOS 1592539510772437 macOS 1592539513190152 macOS 1592539515571793 macOS 1592539516820101 macOS 1592539526219664 macOS 1592539528986263 macOS 1592539534818831 macOS 1592539553012419 macOS 1592539558223361 macOS 1592539561795189 macOS 1592539566529893 macOS 1592539567430571 macOS 1592539568229992 macOS 1592539579712309 macOS 1592539613496863 macOS 1592539619928675 macOS 1592539622314157 macOS 1592539625107522 macOS 1592539630811220 macOS 1592539641260800 macOS 1592539642232052 macOS 1592539648029332 macOS 1592539648209059 macOS 1592539679707233 macOS 1592539687002694 macOS 1592539687702738 macOS 1592539688095638 macOS 1592539691227930 macOS 1592539697637527 macOS 1592539699292247 macOS 1592539706612147 macOS 1592539707026741 macOS 1592539717429416 macOS 1592539721024470 macOS 1592539723035731 macOS 1592539723994355 macOS 1592539726165190 macOS 1592539728016662 macOS 1592539728382483 macOS 1592539736321168 macOS 1592539737729089 macOS 1592539742555450 macOS 1592539756302627 macOS 1592539771811502 macOS 1592539771909095 macOS 1592539775199035 macOS 1592539778120277 macOS 1592539785176603 macOS 1592539788951729 macOS 1592539802173659 macOS 1592539808056692 macOS 1592539814214241 macOS 1592539819815599 macOS 1592539820121493 macOS 1592539825732394 macOS 1592539827290140 macOS 1592539828174968 macOS 1592539839735338 macOS 1592539840167806 macOS 1592539844089540 macOS 1592539844274922 macOS 1592539844673640 macOS 1592539853365400 macOS 1592539853870437 macOS 1592539858122874 macOS 1592539861831406 macOS 1592539862168147 macOS 1592539866460729 macOS 1592539867583317 macOS 1592539877779109 macOS 1592539891458767 macOS 1592539894518057 macOS 1592539896667746 macOS 1592539898915069 macOS 1592539901197116 macOS 1592539901957399 macOS 1592539906595144 macOS 1592539908105402 macOS 1592539908176409 macOS 1592539911365366 macOS 1592539915382610 macOS 1592539918347451 macOS 1592539932058182 macOS 1592539932356253 macOS 1592539944645460 macOS 1592539945686145 macOS 1592539953859025 macOS 1592539960206521 macOS 1592539962751799 macOS 1592539964147207 macOS 1592539965656224 macOS 1592539973265672 macOS 1592539974300508 macOS 1592539977014355 macOS 1592539981200948 macOS 1592539986356712 macOS 1592539989195414 macOS 1592539994568893 macOS 1592540001955600 macOS 1592540002019210 macOS 1592540004839840 macOS 1592540010788765 macOS 1592540019147805 macOS 1592540020091770 macOS 1592540024418823 macOS 1592540025639531 macOS 1592540026505640 macOS 1592540027773277 macOS 1592540030714063 macOS 1592540038646541 macOS 1592540039634736 macOS 1592540047947550 macOS 1592540049672107 macOS 1592540069716718 macOS 1592540072534689 macOS 1592540077867324 macOS 1592540084419487 macOS 159254008787205

-sandbox
%md ##### <img alt="Best Practice" title="Best Practice" style="vertical-align: text-bottom; position: relative; height:1.75em; top:0.3em" src="https://files.training.databricks.com/static/images/icon-blue-ribbon.svg"/> Check your work
- You should only see `macOS` values in the `device` column  
- The fifth row should be an event with timestamp `1592539226602157`

### Classroom Cleanup

In [0]:
%run ./Includes/Classroom-Cleanup

Dropped database and removed files in working directory